internal/c/parts/video/image/image.cpp

```diff
@@ -0,0 +1,408 @@
+        uint32_t r, g, b;
+        uint32_t count;
+    } cubes[256];
+
+    // https://en.wikipedia.org/wiki/Ordered_dithering
+    static uint8_t bayerMatrix[16] = {0, 8, 2, 10, 12, 4, 14, 6, 3, 11, 1, 9, 15, 7, 13, 5};
+
+    IMAGE_DEBUG_PRINT("Converting 32bpp image (%i, %i) to 8bpp", w, h);
+
+    // Allocate memory for new image (8-bit indexed)
+    auto pixels = (uint8_t *)malloc(w * h);
+    if (!pixels) {
+        return nullptr;
+    }
+
+    ZERO_VARIABLE(cubes);
+
+    // Quantization phase
+    auto dst = pixels;
+    for (auto y = 0; y < h; y++) {
+        for (auto x = 0; x < w; x++) {
+            int32_t t = bayerMatrix[((y & 3) << 2) + (x & 3)];
+            int32_t b = image_clamp_component((*src++) + (t << 1));
+            int32_t g = image_clamp_component((*src++) + (t << 1));
+            int32_t r = image_clamp_component((*src++) + (t << 1));
+            ++src; // Ignore alpha
+
+            // Quantize
+            uint8_t k = ((r >> 5) << 5) + ((g >> 5) << 2) + (b >> 6);
+            (*dst++) = k;
+
+            // Prepare RGB cubes for CLUT
+            cubes[k].r += r;
+            cubes[k].g += g;
+            cubes[k].b += b;
+            cubes[k].count++;
+        }
+    }
+
+    // Generate a uniform CLUT based on the quantized colors
+    for (auto i = 0; i < 256; i++) {
+        if (cubes[i].count) {
+            paletteOut[i] = IMAGE_MAKE_BGRA(cubes[i].r / cubes[i].count, cubes[i].g / cubes[i].count, cubes[i].b / cubes[i].count, 0xFF);
+        } else {
+            paletteOut[i] = IMAGE_MAKE_BGRA(0, 0, 0, 0xFF);
+        }
+    }
+
+    return pixels;
+}
+
+/// <summary>
+/// This takes in a 32bpp (BGRA) image raw data and spits out an 8bpp raw image along with it's 256 color (BGRA) palette.
+/// If the number of unique colors in the 32bpp image > 256, then the functions returns a NULL.
+/// Unlike image_convert_8bpp(), no 'real' conversion takes place.
+/// </summary>
+/// <param name="src">The source raw image data. This must be in BGRA format and not NULL</param>
+/// <param name="w">The widht of the image in pixels</param>
+/// <param name="h">The height of the image in pixels</param>
+/// <param name="paletteOut">A 256 color palette if the operation was successful. This cannot be NULL</param>
+/// <returns>A pointer to a 8bpp raw image or NULL if operation failed</returns>
+static uint8_t *image_make_8bpp(uint8_t *src, int w, int h, uint32_t *paletteOut) {
+    IMAGE_DEBUG_PRINT("Extracting 8bpp image (%i, %i) from 32bpp", w, h);
+
+    // Allocate memory for new image (8-bit indexed)
+    auto pixels = (uint8_t *)malloc(w * h);
+    if (!pixels) {
+        return nullptr;
+    }
+
+    int i, j;
+    auto uniqueColors = 0; // As long as this is <= 256 we will keep going until we are done
+    auto src32bpp = (uint32_t *)src;
+    for (i = 0; i < w * h; i++) {
+        auto srcColor = src32bpp[i];
+
+        // Check if the src color exists in our palette
+        for (j = 0; j < uniqueColors; j++) {
```

Comment:
The logic goods good to me, I'm curious if you've seen how it performs with large images with lots of colors. I think it's fine even if it's a bit slower, if necessary I suspect we could optimize this a bit though.

Reply:
Honestly, I did not think of this from a performance standpoint since all this happens only during image load. Also since `uniqueColors` is always `<= 256` I am not sure if it will have any significant slowdowns. Although, it could, if the images are unusually large. I'll run some tests and share here. We can review these numbers before merging the PR.

Reply:
![image](https://user-images.githubusercontent.com/47574584/190865915-51e4f91c-3bd0-4bb1-8aff-85ab9c26cf91.png)
![image](https://user-images.githubusercontent.com/47574584/190865926-1150f0aa-f073-46ae-88ed-f1f9a1206622.png)

lol. We are taking a minute and 2 seconds to load 100 1080p images with all the conversion, compared to 5.8 seconds at 32bpp without any conversion. I am all ears. 😁

Reply:
Ouch :(

I was thinking you could try using a hash table that maps the 32-bit colors to the 8-bit indexes picked for them. That would avoid the `O(n)` comparisons for each pixel by replacing them with a check of whether the 32-bit color is already in the hashmap or not.

Reply:
> Honestly, I did not think of this from a performance standpoint since all this happens only during image load. Also since `uniqueColors` is always `<= 256` I am not sure if it will have any significant slowdowns. Although, it could, if the images are unusually large. I'll run some tests and share here. We can review these numbers before merging the PR.

I would only be using smaller modest 256 color images myself. 320x200 - 1024x768 tops

Reply:
> I would only be using smaller modest 256 color images myself. 320x200 - 1024x768 tops

The good news is that since this PR moved the code into a separate `cpp` file we realized we're able to always build this portion of the code with `-O2` even though we haven't enabled it for the rest of the code. I forget the exact numbers @a740g gave me but I believe after the addition of the hashmap and using `-O2` it only takes 8 seconds to load those images, so significantly better :)

Reply:
That's great. Thanks @mkilgore and @a740g